### ISSUE TITLE
With pagination enabled getAllEntities does not use findAllWithEagerRelationships

### ISSUE THREAD

**Issue by geraldhumphries** [jhipster/generator-jhipster#3603]
##### **Overview of the issue**

I think that this is not a big problem and that the solutions might cause more problems than leaving it be, but I felt it was worth opening an issue for discussion.

Without pagination enabled, `getAllEntities` in entity resources returns a list of objects with owned many-to-many relationships initialized. With pagination, `getAllEntities` returns a list of objects without the owned many-to-many relationships initialized.

Generated code, without pagination:
```
    @RequestMapping(value = "/test-fives",
        method = RequestMethod.GET,
        produces = MediaType.APPLICATION_JSON_VALUE)
    @Timed
    public List<TestFive> getAllTestFives() {
        log.debug("REST request to get all TestFives");
        List<TestFive> testFives = testFiveRepository.findAllWithEagerRelationships();
        return testFives;
    }
```

With pagination:
```
    @RequestMapping(value = "/test-fours",
        method = RequestMethod.GET,
        produces = MediaType.APPLICATION_JSON_VALUE)
    @Timed
    public ResponseEntity<List<TestFour>> getAllTestFours(Pageable pageable)
        throws URISyntaxException {
        log.debug("REST request to get a page of TestFours");
        Page<TestFour> page = testFourRepository.findAll(pageable); 
        HttpHeaders headers = PaginationUtil.generatePaginationHttpHeaders(page, "/api/test-fours");
        return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
    }
```

##### **Motivation for or Use Case** 

This is an inconsistency in generated APIs. However, `findAllWithEagerRelationships` will actually load all entities the paginate it in memory, throwing the warning `firstResult/maxResults specified with collection fetch; applying in memory!`.

##### **JHipster Version(s)** 

master

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory** 

```
{
    "relationships": [
        {
            "relationshipName": "testThree",
            "otherEntityName": "testThree",
            "relationshipType": "many-to-many",
            "otherEntityField": "id",
            "ownerSide": true
        }
    ],
    "fields": [],
    "changelogDate": "20160516151834",
    "dto": "no",
    "service": "no",
    "entityTableName": "test_five",
    "pagination": "no",
    "microserviceName": "q",
    "searchEngine": "elasticsearch"
}
```
```
{
    "relationships": [
        {
            "relationshipName": "testThree",
            "otherEntityName": "testThree",
            "relationshipType": "many-to-many",
            "otherEntityField": "id",
            "ownerSide": true
        }
    ],
    "fields": [],
    "changelogDate": "20160516151712",
    "dto": "no",
    "service": "no",
    "entityTableName": "test_four",
    "pagination": "pagination",
    "microserviceName": "q",
    "searchEngine": "elasticsearch"
}
```

##### **Suggest a Fix** 

Change `List<Entity> findAllWithEagerRelationships();` to `Page<Entity> findAllWithEagerRelationships(Pageable pageable);` then use it in the resource/service instead of the `findAll` method.

I am not sure if that fix is worth applying because Hibernate needs to paginate in-memory which can have pretty severe performance impacts depending on the entity and size of the db. If the results were eagerly loaded or lazily loaded after retrieval, that problem can be avoided.

Another approach (if consistency is more important) is to not use `findAllWithEagerRelationships` for non-paginated entities.

**Comment by jdubois** [comment]
OK, this is mostly a matter of being consistent.

Usually I'm always pushing to have lazy loading, but I think the eager relationship was added because on the main table we display some data from the relationship -> don't you have any lazy initialization error with the pagination code?

This must first be checked, and if the eager relationship isn't needed, let's have the lazy relationship everywhere.

**Comment by geraldhumphries** [comment]
I don't think non-paginated entities display eagerly-loaded data when using getAll, at least by default. The paginated code is straight from the generator.

However I think if using DTOs it may use lazy loading to get the relationship if pagination enabled. Will need to test.

**Comment by jdubois** [comment]
I didn't code this, so this is my understanding of the current code:

- Without pagination: we display the "owned" entity on the list page, so you need to do an "eager" fetch to get it. This is true whether you use DTOs or not. This is a nice feature, and it's quite logical that if you are working on the owner entity, you might want access to the owned entity -> then that's just the default generated code, if you need something more strict you will change the code to stay "lazy" with the "findAll" method.
- With pagination: indeed, doing an eager loading with pagination isn't very efficient. I think this is your point, but anyway you can't do sorting (it's not only pagination, it's "pagination and sorting") on a child entity without loading everything. So, as this is inefficient, we fall back to "lazy" loading the owned entity.

To summarize: we have a nice feature without pagination, and we can't have it with pagination (because of sorting).

So this looks good to me, but this isn't obvious and a better documentation would be great - if someone can confirm what I just wrote it would be great, and then we could add this either in the documentation, or as code comments.
